### ISSUE TITLE
Invalid shared read in test_dropout_layernorm_bwd

### ISSUE THREAD

**Issue by naoyam** [NVIDIA/Fuser#2632]
Branch:
```
commit 0cc22b910b5a6302444850019ae1e4f92974bc83 (HEAD -> main, origin/main, origin/HEAD)
Author: Naoya Maruyama <naoyam@users.noreply.github.com>
Date:   Wed Jul 17 20:49:59 2024 -0700
```

Command:

```
 compute-sanitizer pytest -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688'
```

Results:
```
========= COMPUTE-SANITIZER
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0 -- /opt/conda/pytorch/bin/python
cachedir: .pytest_cache
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/home/nmaruyama/nvfuser/debug1/.hypothesis/examples'))
rootdir: /home/nmaruyama/nvfuser/debug1
plugins: benchmark-4.0.0, hypothesis-6.102.4
collecting ... collected 7716 items / 7713 deselected / 3 selected

benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_nvf_benchmark[dtype=torch.bfloat16-size=[512_18688]] ========= Invalid __shared__ read of size 12 bytes
=========     at <unnamed>::nvfuser_inner_outer_persistent_f0_c1_r0_g0(<unnamed>::Tensor<bool, (int)2, (int)2>, <unnamed>::Tensor<float, (int)1, (int)1>, <unnamed>::Tensor<float, (int)2, (int)2>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)2, (int)2>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)1, (int)1>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)2, (int)2>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)2, (int)2>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)2, (int)2>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)2, (int)2>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)1, (int)1>, <unnamed>::Tensor<<unnamed>::__bfloat, (int)1, (int)1>, <unnamed>::Tensor<float, (int)2, (int)2>, <unnamed>::Tensor<float, (int)2, (int)2>, <unnamed>::Tensor<long long, (int)1, (int)1>)+0x5ad0
=========     by thread (0,1,0) in block (0,0,0)
=========     Address 0xd818 is misaligned
```

**Comment by naoyam** [comment]
@liqiangxl Can you please take a look? The error message seems to indicate a vectorization problem, but I wonder why we issue 12-byte accesses.

**Comment by liqiangxl** [comment]
> @liqiangxl Can you please take a look? The error message seems to indicate a vectorization problem, but I wonder why we issue 12-byte accesses.

I can't reproduce on pjnl-0710 (and the latest 0717), which container were you using?
```
root@nvdl-a112-d004:/opt/pytorch/nvfuser#  compute-sanitizer pytest -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688'
========= COMPUTE-SANITIZER
======================================================================================================================================================== test session starts ========================================================================================================================================================
platform linux -- Python 3.10.12, pytest-8.1.1, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/opt/pytorch/nvfuser/.hypothesis/examples'))
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /opt/pytorch/nvfuser
plugins: timeout-2.3.1, hypothesis-6.104.2, random-order-1.1.1, cov-5.0.0, timestamper-0.0.10, xdist-3.6.1, shard-0.1.2, benchmark-4.0.0
collected 7716 items / 7713 deselected / 3 selected                                                                                                                                                                                                                                                                                 
Running 3 items in this shard: benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_nvf_benchmark[dtype=torch.bfloat16-size=[512_18688]], benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_baseline_benchmark[dtype=torch.bfloat16-size=[512_18688]-eager], benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_baseline_benchmark[dtype=torch.bfloat16-size=[512_18688]-compile]

[2024-07-18 21:01:46] benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_nvf_benchmark[dtype=torch.bfloat16-size=[512_18688]] PASSED                                                                                                                                                                 [ 33%]
[2024-07-18 21:01:48] benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_baseline_benchmark[dtype=torch.bfloat16-size=[512_18688]-eager] SKIPPED (need --benchmark-eager option to run)                                                                                                              [ 66%]
[2024-07-18 21:01:48] benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_baseline_benchmark[dtype=torch.bfloat16-size=[512_18688]-compile] SKIPPED (need --benchmark-torchcompile option to run)                                                                                                     [100%]

========================================================================================================================================================= warnings summary ==========================================================================================================================================================
benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_nvf_benchmark[dtype=torch.bfloat16-size=[512_18688]]
  benchmarks/python/test_dropout_layernorm_bwd.py:141: PytestBenchmarkWarning: Benchmark fixture was not used at all in this test!
    @pytest.mark.parametrize("size", generate_input_sizes(dims=2))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================================================================================================== 1 passed, 2 skipped, 7713 deselected, 1 warning in 3.14s ======================================================================================================================================
[W718 21:01:49.299016405 communicator.cpp:219] Warning: The singleton Communicator isn't available. This is likely because Communicator::cleanup was called more than once or the instance wasn't successfully initialized. (function cleanup)
========= ERROR SUMMARY: 0 errors
```

**Comment by liqiangxl** [comment]
compute-sanitizer
```
root@xxx:/opt/pytorch/nvfuser# compute-sanitizer --version
NVIDIA (R) Compute Sanitizer
Copyright (c) 2020-2024 NVIDIA Corporation
Version 2024.3.0.0 (build 34431801) (public-release)
```

**Comment by naoyam** [comment]
Hmm, it probably depends on device parameters as I don't see this error with smaller sizes. I haven't tried larger sizes than this. My machine has an RTX 6000. Can you try all the problem sizes of the benchmark?

**Comment by naoyam** [comment]
Here's the heuristics parameters:

```
===== Combined InnerOuter Reduction Stats ========
outer_dim_numel: 512
inner_dim_numel: 18688
regs_buffer_size: 242944
smem_buffer_size: 81920
smem_overhead: 2048
vectorize_factor_input: 8
vectorization_factor_tmp_gmem_write: 4
vectorization_factor_outer: 4
multiple_reds_per_blk: 0
threads_per_sm: 256
gdimy: 142
block(64, 4, 1)
===== Reduction Parameters ========
Tag: InnerOuter Register Persistent Heuristic.

Red On Fastest Dim
Persistent Kernel
Project Persistent Buffers
Batches per block: 10

Iteration Domain: blockIdx.y / split grid dimension outer /
Inner Reduction Domain: persistent batch - 10 / vectorize / factor 8
Launch Parameters: BlockDim.x = 64, BlockDim.y = 4, BlockDim.z = -1, GridDim.x = -1, GridDim.y = 142, GridDim.z = -1, Smem Size = 0
Compile Parameters: index_type = int, maxrregcount = 256, enable_magic_zero = 1, enable_ptxas_verbose = 0

====================================


===== Combined InnerOuter Reduction Stats ========
outer_dim_numel: 512
inner_dim_numel: 18688
regs_buffer_size: 242944
smem_buffer_size: 81920
smem_overhead: 2048
vectorize_factor_input: 8
vectorization_factor_tmp_gmem_write: 4
vectorization_factor_outer: 4
multiple_reds_per_blk: 0
threads_per_sm: 256
gdimy: 142
block(64, 4, 1)
===== Reduction Parameters ========
Tag: InnerOuter Register Persistent Heuristic.

Red On Fastest Dim
Persistent Kernel
Project Persistent Buffers
Batches per block: 10

Iteration Domain: blockIdx.y / split grid dimension outer /
Inner Reduction Domain: persistent batch - 10 / vectorize / factor 8
Launch Parameters: BlockDim.x = 64, BlockDim.y = 4, BlockDim.z = -1, GridDim.x = -1, GridDim.y = 142, GridDim.z = -1, Smem Size = 0
Compile Parameters: index_type = int, maxrregcount = 256, enable_magic_zero = 1, enable_ptxas_verbose = 0

====================================

```

**Comment by liqiangxl** [comment]
> Hmm, it probably depends on device parameters as I don't see this error with smaller sizes. I haven't tried larger sizes than this. My machine has an RTX 6000. Can you try all the problem sizes of the benchmark?

Using A100, tried all sizes with `compute-sanitizer pytest -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16'`, captured 2 OOM errors, no memory access error.

`benchmarks/python/test_dropout_layernorm_bwd.py::test_dropout_layernorm_bwd_nvf_benchmark[dtype=torch.bfloat16-size=[16384_36096]] ========= Program hit cudaErrorMemoryAllocation (error 2) due to "out of memory" on CUDA API call to cudaMalloc`

Will try a different card tomorrow.

**Comment by naoyam** [comment]
Yeah, I don't see the error on A100 either, but with the kernel generated on my workstation, I do see the error even with A100. Please try this kernel:

[fail.cu.txt](https://github.com/user-attachments/files/16301873/fail.cu.txt)

```
NVFUSER_EXTERNAL_SRC=fail.cu cs pytest -s -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688' 
```

**Comment by liqiangxl** [comment]
> Yeah, I don't see the error on A100 either, but with the kernel generated on my workstation, I do see the error even with A100. Please try this kernel:
> 
> [fail.cu.txt](https://github.com/user-attachments/files/16301873/fail.cu.txt)
> 
> ```
> NVFUSER_EXTERNAL_SRC=fail.cu cs pytest -s -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688' 
> ```

Still can't reproduce.
I used A100, container from 20240710, main branch code checked out to your posted commit `0cc22b910b5a6302444850019ae1e4f92974bc83` and ran with command:
```
NVFUSER_EXTERNAL_SRC=fail.cu compute-sanitizer pytest -s -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688'
```

I noticed some code diff in the `fail.cu` and the newly generated kernel on `A100`, looks like many runtime files are not updated in `fail.cu`, e.g. there is no `elect_sync` in `fail.cu`, it does't influence `a100` but indicates the outdated runtime file. And many `rand num` related runtime functions are different.

**Comment by naoyam** [comment]
Thanks. Somehow my nvfuser installation was using some old version that I accumulated over the time. Closing as I also don't see the error using the clean container.

**Comment by naoyam** [comment]
@liqiangxl Can you try again this command:

```
compute-sanitizer pytest -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688'
```

I'm seeing a similar misalignment error on one of the A100 nodes.

Here's the generated kernel. This one looks up-to-date.

[fail.cu.txt](https://github.com/user-attachments/files/16318677/fail.cu.txt)

**Comment by liqiangxl** [comment]
> @liqiangxl Can you try again this command:
> 
> ```
> compute-sanitizer pytest -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688'
> ```
> 
> I'm seeing a similar misalignment error on one of the A100 nodes.
> 
> Here's the generated kernel. This one looks up-to-date.
> 
> [fail.cu.txt](https://github.com/user-attachments/files/16318677/fail.cu.txt)

I tried **main branch in pjnl container 20240720 on host nxxx-axxx-d004,** still didn't see error if run with:
`compute-sanitizer pytest -v --disable-benchmarking benchmarks/python/test_dropout_layernorm_bwd.py -k 'bfloat16 and 512_18688'`

The source code you uploaded is different from the one generated, e.g. in `fail.cu.txt` batches per block is 8, while in my local run, it is 10. So I can't use the kernel in `fail.cu.txt`. IIRC, in the initial investigation of this issue, the batches per block is 10 (see [this post](https://github.com/NVIDIA/Fuser/issues/2632#issuecomment-2237687791)) not sure why it changed to 8 in `fail.cu.txt`. codebase is not updated?

**Comment by naoyam** [comment]
Sorry, the error was with `-k "16_14592 and bfloat16"`.

It seems it has something to do with CUDA or compute-sanitizer versions. I was using 12.4. I confirmed the generated kernel is the same with the one generated with the latest PJNL container. The error happens with my own container using 12.4 but not with the PJNL container. I upgraded the CUDA version in my container to 12.5, and it seems the error is gone. 

Closing again for now.
